wetterdienst/dwd/radar/access.py

```diff
@@ -0,0 +1,447 @@
+                & (file_index[DWDMetaColumns.DATETIME.value] < end_date)
+            ]
+
+            if file_index.empty:
+                log.warning(f"No radar file found for {parameter}, {site}, {format}")
+                return
+
+            # Iterate list of files and yield "RadarResult" items.
+            for _, row in file_index.iterrows():
+                date_time = row[DWDMetaColumns.DATETIME.value]
+                url = row[DWDMetaColumns.FILENAME.value]
+
+                for result in _download_generic_data(url=url):
+                    if result.timestamp is None:
+                        result.timestamp = date_time
+                    yield result
+
+
+def should_cache_download(*args, **kwargs) -> bool:  # pragma: no cover
+    """
+    Determine whether this specific result should be cached.
+
+    Here, we don't want to cache any files containing "-latest-" in their filenames.
+
+    :param args: Arguments of decorated function.
+    :param kwargs: Keyword arguments of decorated function.
+    :return: When cache should be dimissed, return False. Otherwise, return True.
+    """
+    url = args[0]
+    if "-latest-" in url:
+        return False
+    return True
+
+
+@payload_cache_five_minutes.cache_on_arguments(should_cache_fn=should_cache_download)
+def _download_generic_data_cached(url: str) -> BytesIO:
+    return url, download_file_from_dwd(url)
+
+
+def _download_generic_data(url: str) -> Generator[RadarResult, None, None]:
+    """
+    Function (cached) that downloads radar data.
+
+    :param url:         The URL to the file on the DWD server
+
+    :return:            The file in binary, either an archive of one file
+                        or an archive of multiple files.
+    """
+
+    _, data = _download_generic_data_cached(url)
+
+    # RadarParameter.FX_REFLECTIVITY
+    if url.endswith(".tar.bz2"):
+        with bz2.BZ2File(data, mode="rb") as archive:
+            with tarfile.open(fileobj=archive) as tar_file:
+                for file in tar_file.getmembers():
+                    yield RadarResult(
+                        data=BytesIO(tar_file.extractfile(file).read()),
+                        timestamp=get_date_from_filename(file.name),
+                        filename=file.name,
+                    )
+
+    # RadarParameter.WN_REFLECTIVITY, RADAR_PARAMETERS_SWEEPS (BUFR)
+    elif url.endswith(".bz2"):
+        with bz2.BZ2File(data, mode="rb") as archive:
+            data = BytesIO(archive.read())
+            yield RadarResult(url=url, data=data, timestamp=get_date_from_filename(url))
+
+    # RADAR_PARAMETERS_RADVOR
+    elif url.endswith(".gz"):
+        with gzip.GzipFile(fileobj=data, mode="rb") as archive:
+            data = BytesIO(archive.read())
+            yield RadarResult(url=url, data=data, timestamp=get_date_from_filename(url))
+
+    else:
+        yield RadarResult(url=url, data=data, timestamp=get_date_from_filename(url))
```

Comment:
After adding support for transparent unzipping of .gz and .bz2 files, some of the `test_api_historic.py` tests apparently fail on Windows with
```
EOFError: Compressed file ended before the end-of-stream marker was reached
```
See also https://github.com/earthobservations/wetterdienst/pull/190/checks?check_run_id=1206522119.

Can I humbly ask one of you to run these on a Windows machine and report about the outcome?

```
pytest -vvvv -k test_api_historic
```

Reply:
@gutzbenj has been able to confirm that:

```
FAILED tests/dwd/radar/test_api_historic.py::test_radar_request_composite_historic_fx_timerange - EOFError: Compressed file ended before the end-of-stream marker was reached
FAILED tests/dwd/radar/test_api_historic.py::test_radar_request_site_historic_sweep_pcp_v_bufr_timerange - EOFError: Compressed file ended before the end-of-stream marker was reached
FAILED tests/dwd/radar/test_api_historic.py::test_radar_request_site_historic_sweep_vol_v_bufr_timerange - EOFError: Compressed file ended before the end-of-stream marker was reached
```

He asked: "Could it be related to caching?"

Reply:
It seems to work flawlessly on Linux and macOS. The failures are coming from the transparent unzipping of .gz and .bz2 files introduced just recently. Damn.

https://github.com/earthobservations/wetterdienst/blob/f576e4607d7088d5e4f76230d4a00218c0b63c72/wetterdienst/dwd/radar/access.py#L216-L252

> Could it be related to caching?

Hm. You might want to test again after commenting out https://github.com/earthobservations/wetterdienst/blob/f576e4607d7088d5e4f76230d4a00218c0b63c72/wetterdienst/dwd/radar/index.py#L45 and https://github.com/earthobservations/wetterdienst/blob/f576e4607d7088d5e4f76230d4a00218c0b63c72/wetterdienst/dwd/radar/access.py#L211.

Reply:
As I am currently clueless about this, I am humbly asking if you can investigate this further on your machine, @gutzbenj. What is noticeable is that all of the failing tests are all `_timerange`-like ones.
- test_radar_request_composite_historic_fx_timerange
- test_radar_request_site_historic_sweep_pcp_v_bufr_timerange
- test_radar_request_site_historic_sweep_vol_v_bufr_timerange

As processing the **gzipped** "radvor" data seems to work flawlessy also on Windows (check `pytest -vvvv -k test_radar_request_radvor`), it seems like this problem is revolving around all **bzip2**'ed artefacts processed with `bz2.BZ2File()`.

Throughout the DWD data repository, these .bz2 files can be found within:

- https://opendata.dwd.de/weather/radar/composit/fx/ (.tar.bz2)
- https://opendata.dwd.de/weather/radar/composit/wn/ (.bz2)
- https://opendata.dwd.de/weather/radar/sites/sweep_pcp_v/boo/ (.bz2)
  Same for `sweep_pcp_z`, `sweep_vol_v` and `sweep_vol_z`.

Reply:
4ae64c4 might have fixed it.

Reply:
One more source might be flaky/not responding internet connection, when the data is processed before the download finished completely. :crossed_fingers: that this is resolved now.